# GREETINGS EVERYONE
For the task of DataVisualization in Tableau we will be using the dataset of Zomato Banglore Restaurants.

You can find the dateset here -> https://www.kaggle.com/code/nkelemen18/starter-zomato-bangalore-restaurants-e22669a4-a
Our goal here is to visualize and extract information based on the past data of orders.

This information can be helpful for upcoming restaurants in Banglore as it will provide critical insights which can help them maximize Profit.

Upon glancing the dataset , we can see that its not as clean as we want so our first task before we can start visualization should be to used the extracted data and transform it for further use

<font color='green'></font>

<font color='green'>Our First step here is to load the dataset and see all the columns available for us to use</font>

In [1]:
import pandas as pd
df = pd.read_csv('zomato.csv')
df.head(5)

url   
0  https://www.zomato.com/bangalore/jalsa-banasha...  \
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name   
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa  \
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone   
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233  \
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type   
0  Banashankari        Casual Dining  \
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked   
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...  \
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)   
0  North Indian, Mughlai, Chinese                         800  \
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item   
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []  \
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [2]:
df.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost(for two people)', 'reviews_list', 'menu_item',
       'listed_in(type)', 'listed_in(city)'],
      dtype='object')

In [3]:
df.shape

(51717, 17)

<font color='green'> We can see we have data for more than 50,000 orders and we have 17 rows to work with.

<font color='green'>Next we will analyse individual columns to further see if data is of any use to us


In [4]:
print('Number of Unique Values')
print(df.nunique())

Number of Unique Values
url                            51717
address                        11495
name                            8792
online_order                       2
book_table                         2
rate                              64
votes                           2328
phone                          14926
location                          93
rest_type                         93
dish_liked                      5271
cuisines                        2723
approx_cost(for two people)       70
reviews_list                   22513
menu_item                       9098
listed_in(type)                    7
listed_in(city)                   30
dtype: int64


<font color='green'>On First glance columns like url, phone, reviews_list are really not of our use since they have way too many categorical values that cannot be generalized. Even column address since we have column listed_in(city) which captures the area we can afford to remove these columns.
</font>

In [5]:
drop_column = ['url','address','phone','reviews_list']
for i in drop_column:
    df.drop(i,axis=1,inplace=True)

<font color='green'>Now we can check for miscellaneous data , such as null values , or that online_order has only 2 unique values , rating is given only out of 5 </font>

In [6]:
df.isna().sum()

name                               0
online_order                       0
book_table                         0
rate                            7775
votes                              0
location                          21
rest_type                        227
dish_liked                     28078
cuisines                          45
approx_cost(for two people)      346
menu_item                          0
listed_in(type)                    0
listed_in(city)                    0
dtype: int64

<font color='green'>Firstly we will delete all 21 rows where location is NUll & approx_cost column also, it makes no sense to assume that</font>

In [7]:
df['location'].dropna(how=any,axis=0,inplace=True)

In [8]:
df['approx_cost(for two people)'].dropna(how=any,axis=0,inplace=True)

In [9]:
df['online_order'].value_counts()

online_order
Yes    30444
No     21273
Name: count, dtype: int64

In [10]:
df['book_table'].value_counts()

book_table
No     45268
Yes     6449
Name: count, dtype: int64

In [11]:
df['votes'].value_counts()

votes
0       10027
4        1140
6         992
7         872
9         738
        ...  
3673        1
1862        1
3909        1
2155        1
843         1
Name: count, Length: 2328, dtype: int64

In [12]:
df['rate'].value_counts()

rate
NEW       2208
3.9/5     2098
3.8/5     2022
3.7/5     2011
3.9 /5    1874
          ... 
2.2 /5       7
2.0 /5       7
2.0/5        4
1.8 /5       3
1.8/5        2
Name: count, Length: 64, dtype: int64

In [13]:
df[df['rate'].isna()].head(2)

name online_order book_table rate  votes      location   
84       Chill Out           No         No  NaN      0  Banashankari  \
90  Me And My Cake           No         No  NaN      0  Banashankari   

      rest_type dish_liked               cuisines approx_cost(for two people)   
84  Quick Bites        NaN  South Indian, Chinese                         100  \
90     Delivery        NaN       Bakery, Desserts                         500   

   menu_item listed_in(type) listed_in(city)  
84        []        Delivery    Banashankari  
90        []        Delivery    Banashankari

<font color='green'>We can see that rate column has a lot of rows as NaN ,  basically either a rating was not given or it was not recorded .  We will transform all Null values to 'Not Reported' so as to not lose any information</font>

In [14]:
df.loc[df['rate'].isna(),'rate'] = 'Not Rated'

In [15]:
df['rest_type'].value_counts()

rest_type
Quick Bites                   19132
Casual Dining                 10330
Cafe                           3732
Delivery                       2604
Dessert Parlor                 2263
                              ...  
Dessert Parlor, Kiosk             2
Food Court, Beverage Shop         2
Dessert Parlor, Food Court        2
Sweet Shop, Dessert Parlor        1
Quick Bites, Kiosk                1
Name: count, Length: 93, dtype: int64

<font color='green'>Since we dont know the the Type of Restaurant for and the number of records for the same is also low we shall remove the rows with missing values

In [16]:
df['rest_type'].dropna(how='any',axis=0,inplace=True)

<font color='green'>Now, we Can see that a restaurant can have two types together as comma seperated, since this comma seperated types is not of much use to us as during visualization it will be seen as a unique entity , we shall divide it , the approach we shall use is similar to one hot encoding . i.e for each unique type of rest we will have a column, then for every row we shall check the type and whatever type is mentioned be it single or comma seperated value , ww will give a values of 1 in the column for that type

In [17]:
unique_rest_type = set()

for rest in df['rest_type']:
    rest_str = str(rest)
    unique_rest_type.update(rest_str.split(', '))

In [18]:
unique_rest_type

{'Bakery',
 'Bar',
 'Beverage Shop',
 'Bhojanalya',
 'Cafe',
 'Casual Dining',
 'Club',
 'Confectionery',
 'Delivery',
 'Dessert Parlor',
 'Dhaba',
 'Fine Dining',
 'Food Court',
 'Food Truck',
 'Irani Cafee',
 'Kiosk',
 'Lounge',
 'Meat Shop',
 'Mess',
 'Microbrewery',
 'Pop Up',
 'Pub',
 'Quick Bites',
 'Sweet Shop',
 'Takeaway',
 'nan'}

<font color='green'> Now that we have all the unique values of rest_type we will create the columns for them

In [19]:
for rest in unique_rest_type:
    df['rest_type_'+rest]=0

In [20]:
df.head(2)

name online_order book_table   rate  votes      location   
0           Jalsa          Yes        Yes  4.1/5    775  Banashankari  \
1  Spice Elephant          Yes         No  4.1/5    787  Banashankari   

       rest_type                                         dish_liked   
0  Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...  \
1  Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   

                         cuisines approx_cost(for two people)  ...   
0  North Indian, Mughlai, Chinese                         800  ...  \
1     Chinese, North Indian, Thai                         800  ...   

  rest_type_Kiosk rest_type_Irani Cafee rest_type_Pop Up   
0               0                     0                0  \
1               0                     0                0   

   rest_type_Food Court  rest_type_Beverage Shop  rest_type_Dhaba   
0                     0                        0                0  \
1                     0                        0                0   

   rest_type_Delivery  rest_type_Bakery  rest_type_nan  rest_type_Bar  
0                   0                 0              0              0  
1                   0                 0              0              0  

[2 rows x 39 columns]

<font color='green'> Now we shall enumerate over all the rows get the row id and the rest_type and based on the value of rest_type assign 1 to the respective columns

In [21]:
for i,rest in enumerate(df['rest_type']):
    rest_str = str(rest)
    for x in rest_str.split(', '):
        df.at[i,'rest_type_'+x]=1

In [22]:
df[df['rest_type_Dhaba']==1]

name online_order book_table       rate  votes   
1001           Sri Punjabi Dhaba          Yes         No      3.3/5     39  \
2878                    PK Dhaba          Yes         No        NEW      0   
7323                    PK Dhaba          Yes         No        NEW      0   
8125                    PK Dhaba          Yes         No        NEW      0   
8486           Sri Punjabi Dhaba          Yes         No      3.2/5     44   
9061                    PK Dhaba          Yes         No        NEW      0   
10170    Harilal ka Bihari Dhaba          Yes         No        NEW      0   
10539          Sri Punjabi Dhaba          Yes         No      3.2/5     44   
16750    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
19591          Sri Punjabi Dhaba          Yes         No      3.3/5     47   
20192                   PK Dhaba          Yes         No        NEW      0   
21726          Sri Punjabi Dhaba          Yes         No     3.3 /5     50   
21928          Sri Punjabi Dhaba          Yes         No      3.3/5     46   
22415                   PK Dhaba          Yes         No        NEW      0   
23820          Sri Punjabi Dhaba          Yes         No      3.3/5     46   
23963          Jai Bhavani Dhaba          Yes         No      3.0/5     52   
25267          Jai Bhavani Dhaba          Yes         No      3.0/5     52   
26739          Sri Punjabi Dhaba          Yes         No      3.3/5     49   
27800    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
28355  Jee Aaya Nu Punjabi Dhaba           No         No  Not Rated      0   
28635    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
29454          Sri Punjabi Dhaba          Yes         No     3.3 /5     50   
30861    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
31419  Jee Aaya Nu Punjabi Dhaba           No         No  Not Rated      0   
31424    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
32355          Sri Punjabi Dhaba          Yes         No     3.3 /5     50   
33617    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
34149    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
34154  Jee Aaya Nu Punjabi Dhaba           No         No  Not Rated      0   
34988          Sri Punjabi Dhaba          Yes         No     3.3 /5     50   
35627                   PK Dhaba          Yes         No  Not Rated      0   
36378    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
36885  Jee Aaya Nu Punjabi Dhaba           No         No  Not Rated      0   
37272    Harilal ka Bihari Dhaba          Yes         No  Not Rated      0   
37536          Sri Punjabi Dhaba          Yes         No     3.3 /5     50   
39198              Punjabi Dhaba           No         No     3.6 /5     79   
41021                   PK Dhaba          Yes         No  Not Rated      0   
41706                   PK Dhaba          Yes         No  Not Rated      0   
51277      Om Sai Punjabi Dhabha          Yes         No     3.4 /5     10   
51278                  Aishwarya           No         No     3.3 /5      6   

                    location rest_type         dish_liked   
1001                     BTM     Dhaba                NaN  \
2878            Marathahalli     Dhaba                NaN   
7323            Marathahalli     Dhaba                NaN   
8125            Marathahalli     Dhaba                NaN   
8486                     BTM     Dhaba                NaN   
9061            Marathahalli     Dhaba                NaN   
10170  Koramangala 1st Block     Dhaba                NaN   
10539                    BTM     Dhaba                NaN   
16750  Koramangala 1st Block     Dhaba                NaN   
19591                    BTM     Dhaba                NaN   
20192           Marathahalli     Dhaba                NaN   
21726                    BTM     Dhaba                NaN   
21928                    BTM     Dha

<font color = 'green'>In this example we can see that when rest_type == Dhaba our new column rest_typeClub has a value of 1, GREAT GOOD JOB!! One Step Closer  
    
<font color = 'green'>Moving forward to dish_liked 

In [23]:
df['dish_liked'].isna().sum()

28078

<font color = 'green'> With around 28,000 rows of Null value it is important to look at this , basically we can assume that no dish was liked at the restuarant , so we shall convert all Null values to 'None'

In [24]:
df['dish_liked'].fillna('None',inplace=True)

In [25]:
df['dish_liked'].value_counts()

dish_liked
None                                                                                                 28078
Biryani                                                                                                182
Chicken Biryani                                                                                         73
Friendly Staff                                                                                          69
Waffles                                                                                                 68
                                                                                                     ...  
Butter Chicken, Shawarma Roll, Chicken Shawarama, Chicken Grill, Rolls, Al Faham Chicken, Biryani        1
Filter Coffee, Sandwich, Bonda, Vada, Masala Dosa, Salad, Aloo Curry                                     1
Burgers, Fries, Jumbo Royale Burger, Salads, Peri Peri Chicken Salad, Potato Wedges, Rolls               1
Chaat, Pav Bhaji, Raj Kach

<font color='green'> With this column the same problem arises of dishes being mentioned as comma seperated so we shall perform the same task of creating diff columns for each dish as a whole

In [26]:
unique_dishes = set ()

for i in df['dish_liked']:
    dish_str = str(i)
    unique_dishes.update(dish_str.split(', '))

In [27]:
len(unique_dishes)

2793

In [28]:
unique_dishes

{'Palak Chicken',
 'Galauti Kebab',
 'Veg Mushroom',
 'Mutton Fry',
 'Chicken Chatpata',
 'Espresso',
 'Gur Ice Cream',
 'Doner Kebabs',
 'Buffalo Wings',
 'Sweet Pongal',
 'Nalli Nihari',
 'Masala Buttermilk',
 'Kadhai Mushroom',
 'Dum Ka Chicken',
 'Mughlai Chicken',
 'Cheese Dabeli',
 'Chilli Bhajji',
 'Mediterranean Platter',
 'Florentine Omelette',
 'Coconut Rice',
 'Chicken Drumsticks',
 'Baingan Bharta',
 'Breads',
 'Chicken Teriyaki Bao',
 'Truffle',
 'Mixed Fried Rice',
 'Benne Khali Dosa',
 'Orange Mousse',
 'Gong Bao Chicken',
 'Brownie Fudge',
 'Mushroom Bruschettas',
 'Paneer Roll',
 'Cafe Mocha',
 'Bun Nippat Cheese',
 'Pallipalayam Chicken',
 'Oreo Delight',
 'Pulled Pork Burger',
 'Chocolate Fondue',
 'American Chopsuey',
 'Tomato Chicken',
 'Garlic Maggi',
 'King Prawns',
 'Puttu Biryani',
 'Mexican Platter',
 'Mango Roll',
 'Rasam Rice',
 'Hummus Pita',
 'Chicken Parantha',
 'Chocolate Chip Waffle',
 'Idli',
 'Tamatar Ka Shorba',
 'Cheese Maggi',
 'Masala Tea',
 'Stuf

<font color='Green'>It makes absolutely no sense to compute additional 2793 rows for visualization , so we shall create a temporary df in which we will create these one hot encoding columns and we will sort them based on count and take the number of columns that make most sense

In [30]:
df_dishes = df['dish_liked'].str.get_dummies(sep=',')

In [31]:
df_dishes

Aalo Bhaja   Aam Panna   Aamras   Abbabi Chicken   Achari Chicken   
0                0           0        0                0                0  \
1                0           0        0                0                0   
2                0           0        0                0                0   
3                0           0        0                0                0   
4                0           0        0                0                0   
...            ...         ...      ...              ...              ...   
51712            0           0        0                0                0   
51713            0           0        0                0                0   
51714            0           0        0                0                0   
51715            0           0        0                0                0   
51716            0           0        0                0                0   

        Achari Mushroom   Achari Paneer   Adrak Ke Panje   Afghan Chicken   
0                     0               0                0                0  \
1                     0               0                0                0   
2                     0               0                0                0   
3                     0               0                0                0   
4                     0               0                0                0   
...                 ...             ...              ...              ...   
51712                 0               0                0                0   
51713                 0               0                0                0   
51714                 0               0                0                0   
51715                 0               0                0                0   
51716                 0               0                0                0   

        Afghani Biryani  ...  Watermelon Juice  Wedges  Wheat Beer   
0                     0  ...                 0       0           0  \
1                     0  ...                 0       0           0   
2                     0  ...                 0       0           0   
3                     0  ...                 0       0           0   
4                     0  ...                 0       0           0   
...                 ...  ...               ...     ...         ...   
51712                 0  ...                 0       0           0   
51713                 0  ...                 0       0           0   
51714                 0  ...                 0       0           0   
51715                 0  ...                 0       0           0   
51716                 0  ...                 0       0           0   

       White Pasta  Whole Meat Shawarma  Wine  Wings  Wonton  Wood Fire Pizza   
0                0                    0     0      0       0                0  \
1                0                    0     0      0       0                0   
2                0                    0     0      0       0                0   
3                0                    0     0      0       0                0   
4                0                    0     0      0       0                0   
...            ...                  ...   ...    ...     ...              ...   
51712            0                    0     0      0       0                0   
51713            0                    0     0      0       0                0   
51714            0                    0     0      0       0                0   
51715            0                    0     0      0       0                0   
51716            0                    0     0      0       0                0   

       Zinger Burger  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
...              ...  
51712              0  
51713              0  
51714              0  
51715              0  
51716              0  

[51717 rows x 3514 columns]

<Font color='green'>Now we have the columns ready and we shall pick the top 51 columns where the sum is largest which basically means the the dishes which were ordered the most and concat to our main df , further we should remove the joint dish_liked column from our df 

In [32]:
top_dishes = df_dishes.sum().nlargest(51)
df_top_dishes = df_dishes[top_dishes.index]
df = pd.concat([df, df_top_dishes], axis=1)

# Drop the original "dish_liked" column if needed
df.drop('dish_liked', axis=1, inplace=True)
df.drop('None', axis=1, inplace=True)

In [33]:
df.drop('cuisines', axis=1, inplace=True)
df.drop('menu_item', axis=1, inplace=True)

In [34]:
df.to_csv('zomato_transform.csv',index=False)